### PR TITLE
feat: `jenkins` plugin prod ready

### PR DESCRIPTION
**Pull request by aFlyBird0**
## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests

## Description

1. support test mod and prod mode meanwhile.
* test: auto-create PV named `jenkins-pv` and use host path
* prod: user can customize `storageClass`
2. use NodePort to expose service.
3. support `nodePort` config customize.
4. show Jenkins URL and tell user how to get the password of user admin
5. docs improved.

## Related Issues
#758 

### 1. dtm apply
 ./dtm apply -f jenkins.yaml -y
2022-06-27 20:48:26 ℹ [INFO]  Apply started.
2022-06-27 20:48:26 ℹ [INFO]  Got Backend from config: local
2022-06-27 20:48:26 ℹ [INFO]  Using dir <.devstream> to store plugins.
2022-06-27 20:48:27 ℹ [INFO]  Using local backend. State file: devstream.state.
2022-06-27 20:48:27 ℹ [INFO]  Tool (jenkins/default) found in config but doesn't exist in the state, will be created.
2022-06-27 20:48:27 ℹ [INFO]  Start executing the plan.
2022-06-27 20:48:27 ℹ [INFO]  Changes count: 1.
2022-06-27 20:48:27 ℹ [INFO]  -------------------- [  Processing progress: 1/1.  ] --------------------
2022-06-27 20:48:27 ℹ [INFO]  Processing: (jenkins/default) -> Create ...
2022-06-27 20:48:27 ℹ [INFO]  Test environment is enabled. Please ensure you have created the directories correctly under the guide of plugin doc.
2022-06-27 20:48:27 !! [ERROR]  Failed to create PersistentVolume < jenkins-pv >: persistentvolumes "jenkins-pv" already exists.
2022-06-27 20:48:27 ℹ [INFO]  The resource PersistentVolume is already exists.
2022-06-27 20:48:28 ℹ [INFO]  Creating or updating helm chart ...
2022/06/27 20:48:31 creating 11 resource(s)
2022/06/27 20:48:31 beginning wait for 11 resources with timeout of 5m0s
2022/06/27 20:48:31 StatefulSet is not ready: jenkins/dev-jenkins. 0 out of 1 expected pods are ready
2022/06/27 20:48:33 StatefulSet is not ready: jenkins/dev-jenkins. 0 out of 1 expected pods are ready
...
...
2022/06/27 20:51:47 StatefulSet is not ready: jenkins/dev-jenkins. 0 out of 1 expected pods are ready
2022/06/27 20:51:49 StatefulSet is not ready: jenkins/dev-jenkins. 0 out of 1 expected pods are ready
2022/06/27 20:51:51 StatefulSet is not ready: jenkins/dev-jenkins. 0 out of 1 expected pods are ready
2022/06/27 20:51:53 release installed successfully: dev/jenkins-4.1.11
2022-06-27 20:51:53 ℹ [INFO]  Here is how to get the password of the admin user:
2022-06-27 20:51:53 ℹ [INFO]  kubectl exec --namespace jenkins -it svc/dev-jenkins -c jenkins -- /bin/cat /run/secrets/additional/chart-admin-password && echo
2022-06-27 20:51:53 ℹ [INFO]  Jenkins url:http://172.18.0.2:32000/login

2022-06-27 20:51:53 ✔ [SUCCESS]  Tool (jenkins/default) Create done.
2022-06-27 20:51:53 ℹ [INFO]  -------------------- [  Processing done.  ] --------------------
2022-06-27 20:51:53 ✔ [SUCCESS]  All plugins applied successfully.
2022-06-27 20:51:53 ✔ [SUCCESS]  Apply finished.

(Note: `persistentvolumes "jenkins-pv" already exists.` can be ignored.)

image

image

image

### 2. dtm delete

❯ ./dtm delete -f jenkins.yaml -y
2022-06-27 20:59:19 ℹ [INFO]  Delete started.
2022-06-27 20:59:19 ℹ [INFO]  Got Backend from config: local
2022-06-27 20:59:19 ℹ [INFO]  Using dir <.devstream> to store plugins.
2022-06-27 20:59:19 ℹ [INFO]  Using local backend. State file: devstream.state.
2022-06-27 20:59:19 ℹ [INFO]  Tool (jenkins/default) will be deleted.
2022-06-27 20:59:19 ℹ [INFO]  Start executing the plan.
2022-06-27 20:59:19 ℹ [INFO]  Changes count: 1.
2022-06-27 20:59:19 ℹ [INFO]  -------------------- [  Processing progress: 1/1.  ] --------------------
2022-06-27 20:59:19 ℹ [INFO]  Processing: (jenkins/default) -> Delete ...
2022-06-27 20:59:21 ℹ [INFO]  Uninstalling jenkins helm chart ...
2022/06/27 20:59:21 uninstall: Deleting dev
2022/06/27 20:59:23 Starting delete for "dev-jenkins" Service
2022/06/27 20:59:23 Starting delete for "dev-jenkins-agent" Service
2022/06/27 20:59:23 Starting delete for "dev-jenkins" StatefulSet
2022/06/27 20:59:23 Starting delete for "dev-jenkins-watch-configmaps" RoleBinding
2022/06/27 20:59:23 Starting delete for "dev-jenkins-schedule-agents" RoleBinding
2022/06/27 20:59:23 Starting delete for "dev-jenkins-schedule-agents" Role
2022/06/27 20:59:23 Starting delete for "dev-jenkins-casc-reload" Role
2022/06/27 20:59:23 Starting delete for "dev-jenkins" PersistentVolumeClaim
2022/06/27 20:59:23 Starting delete for "dev-jenkins-jenkins-jcasc-config" ConfigMap
2022/06/27 20:59:23 Starting delete for "dev-jenkins" ConfigMap
2022/06/27 20:59:23 Starting delete for "dev-jenkins" Secret
2022/06/27 20:59:23 purge requested for dev
2022/06/27 20:59:23 release uninstalled, response: &{0x14000631a40 }
2022-06-27 20:59:23 ⚠ [WARN]  

NOTICE!!!
The PersistentVolume jenkins-pv is NOT been deleted.
You can execute the "kubectl delete pv jenkins-pv" to delete it manually.The local data directory /Users/lhp/data/jenkins-volumes is also NOT been deleted.You can delete it manually.

2022-06-27 20:59:23 ℹ [INFO]  Prepare to delete 'jenkins_default' from States.
2022-06-27 20:59:23 ✔ [SUCCESS]  Tool (jenkins/default) delete done.
2022-06-27 20:59:23 ℹ [INFO]  -------------------- [  Processing done.  ] --------------------
2022-06-27 20:59:23 ✔ [SUCCESS]  All plugins deleted successfully.
2022-06-27 20:59:23 ✔ [SUCCESS]  Delete finished.

### CODE REVIEW

File: docs/plugins/jenkins.md [hunk 1]
```diff
@@ -4,15 +4,36 @@ This plugin installs [jenkins](https://jenkins.io) in an existing Kubernetes clu
 
 ## Usage
 
-NOTICE:
+### NOTICE
 
-Create the data directory manually:
+#### Production Environments
+Please be sure to change the `storageClass` in the configuration item to an existing storage class.
```
Comment (daniel-hutao) [inline docs/plugins/jenkins.md:10]: storage class -> StorageClass

File: docs/plugins/jenkins.md [hunk 2]
```diff
@@ -4,15 +4,36 @@ This plugin installs [jenkins](https://jenkins.io) in an existing Kubernetes clu
...
+2. Create the data directory manually in the host where kubernetes is running:
 
 ```bash
+# If you run k8s and dtm in the same host
 mkdir -p ~/data/jenkins-volumes/
 chown -R 1000:1000 ~/data/jenkins-volumes/
+
+-------------------------
+
+# If you run k8s and dtm in different hosts, such as run k8s in docker
```
Comment (daniel-hutao) [inline docs/plugins/jenkins.md:25]: It's better to tell users how to execute `chown` with Minikube. e.g. `minikube ssh & mkdir ... `

File: docs/plugins/jenkins.zh.md [hunk 3]
```diff
@@ -1,3 +1,43 @@
...
+# 1. 获取 dtm 运行的主机的用户的 home 目录
+cd ~ && pwd
+# 2. 复制上面的命令结果
+# 3. 在 k8s 运行的主机上创建数据目录并修改权限，命令如下：
+docker exec -it <container-id> bash
+mkdir -p <your-dtm-home-dir>/data/jenkins-volumes/
+chown -R 1000:1000 <your-dtm-home-dir>/data/jenkins-volumes/
+```
+
+
```
Comment (daniel-hutao) [inline docs/plugins/jenkins.zh.md:35]: Redundant blank lines

File: internal/pkg/plugin/jenkins/create.go [hunk 4]
```diff
@@ -61,5 +61,11 @@ func Create(options map[string]interface{}) (map[string]interface{}, error) {
 	retMap := GetStaticState(releaseName).ToStringInterfaceMap()
 	log.Debugf("Return map: %v.", retMap)
 
+	// show how to get pwd of the admin user
+	howToGetPwdOfAdmin(&opts)
```
Comment (daniel-hutao) [inline internal/pkg/plugin/jenkins/create.go:65]: It's rarely to use `pwd` for password. Maybe `passwd` is more readable.

File: internal/pkg/plugin/jenkins/options.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package jenkins
+
+import "github.com/devstream-io/devstream/internal/pkg/plugin/common/helm"
+
+type Options struct {
+	// in test env, we will use hostpath to auto create persistent volume
+	TestEnv      bool `mapstructure:"test_env"`
+	helm.Options `mapstructure:",squash"`
```
Comment (daniel-hutao) [inline internal/pkg/plugin/jenkins/options.go:8]: very good!

File: internal/pkg/plugin/jenkins/precreate.go [hunk 6]
```diff
@@ -16,29 +14,28 @@ import (
...
 
 const (
-	JenkinsName          = "jenkins"
-	JenkinsNamespace     = "jenkins"
-	JenkinsDataDirectory = "/data/jenkins-volume/"
-	JenkinsUid           = 1000
-	JenkinsGid           = 1000
+	JenkinsName                      = "jenkins"
+	JenkinsNamespace                 = "jenkins"
+	JenkinsDataDirectory             = "/data/jenkins-volumes/"
```
Comment (daniel-hutao) [inline internal/pkg/plugin/jenkins/precreate.go:19]: `/data/jenkins-volume` is better. It's a single volume.

**Review by daniel-hutao — CHANGES_REQUESTED** [review]

File: internal/pkg/plugin/jenkins/precreate.go (hunk 6 above)
Comment (aFlyBird0) [reply internal/pkg/plugin/jenkins/precreate.go:19]: This is more precise, I agree with you.

File: docs/plugins/jenkins.md (hunk 2 above)
Comment (aFlyBird0) [reply docs/plugins/jenkins.md:25]: Thank you very much! ❤️

**Review by daniel-hutao — APPROVED** [review]

File: docs/plugins/jenkins.md [hunk 7]
```diff
@@ -4,15 +4,40 @@ This plugin installs [jenkins](https://jenkins.io) in an existing Kubernetes clu
 
 ## Usage
 
-NOTICE:
+### NOTICE
 
-Create the data directory manually:
+#### Production Environments
```
Comment (IronCore864) [inline docs/plugins/jenkins.md:9]: Level 4 titles should reside under level 3 titles. Please adjust the title levels and doc structure. Same for below titles.

**Review by IronCore864 — CHANGES_REQUESTED** [review]
Also I've got a few issues which have been communicated over IM.

**Review by IronCore864 — APPROVED** [review]

**Pull request merged**